samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java

```diff
@@ -71,54 +63,29 @@
 
   @Override
   protected void handleInit(Config config, TaskContext context) {
-    keysRemoved = context.getMetricsRegistry()
-        .newCounter(OperatorImpl.class.getName(), getOperatorName() + "-keys-removed");
     this.thisPartialJoinFn.init(config, context);
   }
 
   @Override
-  public Collection<RM> handleMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
+  public Collection<JM> handleMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
     K key = thisPartialJoinFn.getKey(message);
-    thisPartialJoinFn.getState().put(key, new PartialJoinMessage<>(message, clock.currentTimeMillis()));
-    PartialJoinMessage<JM> otherMessage = otherPartialJoinFn.getState().get(key);
+    thisPartialJoinFn.getState().put(key, new TimestampedValue<>(message, clock.currentTimeMillis()));
+    TimestampedValue<OM> otherMessage = otherPartialJoinFn.getState().get(key);
     long now = clock.currentTimeMillis();
-    if (otherMessage != null && otherMessage.getReceivedTimeMs() > now - ttlMs) {
-      RM joinResult = thisPartialJoinFn.apply(message, otherMessage.getMessage());
+    if (otherMessage != null && otherMessage.getTimestamp() > now - ttlMs) {
+      JM joinResult = thisPartialJoinFn.apply(message, otherMessage.getValue());
       return Collections.singletonList(joinResult);
     }
     return Collections.emptyList();
   }
 
   @Override
```

Comment:
Do we rely on rocksDb ttl for expiry? Instead of doing point-deletes on the store?

Reply:
Yeah, we configure RocksDb TTL  + changelog TTL == ttlMs.